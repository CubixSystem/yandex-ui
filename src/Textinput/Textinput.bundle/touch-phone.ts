import { compose, composeU, ExtractProps } from '@bem-react/core';

import {
    Textinput as TextinputTouchPhone,
} from '../Textinput@touch-phone';
import { withDebounceInput } from '../../withDebounceInput';
// _size
import { withSizeM } from '../_size/Textinput_size_m';
import { withSizeS } from '../_size/Textinput_size_s';
// _theme
import { withThemeNormal } from '../_theme/Textinput_theme_normal';
import { withThemeWebsearch } from '../_theme/Textinput_theme_websearch';
// _view
import { withViewDefault } from '../_view/Textinput_view_default';
import { withViewMaterial } from '../_view/Textinput_view_material';
// _hasClear
import { withHasClear } from '../_hasClear/Textinput_hasClear@touch-phone';
// _pin
import { withPinBrickBrick } from '../_pin/Textinput_pin_brick-brick';
import { withPinBrickClear } from '../_pin/Textinput_pin_brick-clear';
import { withPinBrickRound } from '../_pin/Textinput_pin_brick-round';
import { withPinClearBrick } from '../_pin/Textinput_pin_clear-brick';
import { withPinClearClear } from '../_pin/Textinput_pin_clear-clear';
import { withPinClearRound } from '../_pin/Textinput_pin_clear-round';
import { withPinRoundBrick } from '../_pin/Textinput_pin_round-brick';
import { withPinRoundClear } from '../_pin/Textinput_pin_round-clear';
import { withPinRoundRound } from '../_pin/Textinput_pin_round-round';
// _baseline
import { withBaseline } from '../_baseline/Textinput_baseline';

export * from '../Textinput@touch-phone';

export const Textinput = compose(
    withDebounceInput,
    composeU(withSizeM, withSizeS),
    composeU(withThemeNormal, withThemeWebsearch),
    composeU(
        composeU(
            withPinBrickBrick,
            withPinBrickClear,
            withPinBrickRound,
            withPinClearBrick,
            withPinClearClear,
            withPinClearRound,
        ),
        withPinRoundBrick,
        withPinRoundClear,
        withPinRoundRound,
    ),
    composeU(
        withViewDefault,
        withViewMaterial,
    ),
    withBaseline,
    withHasClear,
)(TextinputTouchPhone);

export type ITextinputProps = ExtractProps<typeof Textinput>;
