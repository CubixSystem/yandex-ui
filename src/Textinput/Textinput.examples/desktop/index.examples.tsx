import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export * from './baseline.examples';
export * from './debounce.examples';
export * from './has-clear.examples';
export * from './pin.examples';
export * from './size.examples';
export * from './state.examples';
export * from './theme.examples';
export * from './type.examples';
export * from './view.examples';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: createDecorators({ scope: 'desktop' }),
    parameters,
};
