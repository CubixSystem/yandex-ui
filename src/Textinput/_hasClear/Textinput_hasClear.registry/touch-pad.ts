import { TextinputClear } from '../../Clear/Textinput-Clear';
import { textinputRegistry as textinputRegistryBase } from '../../Textinput.registry/touch-pad';

export const textinputRegistry = textinputRegistryBase.set('Clear', TextinputClear);
