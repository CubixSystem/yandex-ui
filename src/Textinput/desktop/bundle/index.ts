import { compose, composeU, ExtractProps } from '@bem-react/core';

import { Textinput as TextinputDesktop } from '../../Textinput@desktop';
import { withDebounceInput } from '../../../withDebounceInput';
// _size
import { withSizeM } from '../../_size/Textinput_size_m';
import { withSizeS } from '../../_size/Textinput_size_s';
// _theme
import { withThemeNormal } from '../../_theme/Textinput_theme_normal@desktop';
import { withThemeWebsearch } from '../../_theme/Textinput_theme_websearch@desktop';
// _view
import { withViewDefault } from '../../_view/Textinput_view_default@desktop';
import { withViewMaterial } from '../../_view/Textinput_view_material@desktop';
// _hasClear
import { withHasClear } from '../../_hasClear/Textinput_hasClear@desktop';
// _pin
import { withPinBrickBrick } from '../../_pin/Textinput_pin_brick-brick';
import { withPinBrickClear } from '../../_pin/Textinput_pin_brick-clear';
import { withPinBrickRound } from '../../_pin/Textinput_pin_brick-round';
import { withPinClearBrick } from '../../_pin/Textinput_pin_clear-brick';
import { withPinClearClear } from '../../_pin/Textinput_pin_clear-clear';
import { withPinClearRound } from '../../_pin/Textinput_pin_clear-round';
import { withPinRoundBrick } from '../../_pin/Textinput_pin_round-brick';
import { withPinRoundClear } from '../../_pin/Textinput_pin_round-clear';
import { withPinRoundRound } from '../../_pin/Textinput_pin_round-round';
// _baseline
import { withBaseline } from '../../_baseline/Textinput_baseline';

export * from '../../Textinput@desktop';

export const Textinput = compose(
    withDebounceInput,
    composeU(withSizeM, withSizeS),
    composeU(withThemeNormal, withThemeWebsearch),
    composeU(
        composeU(
            withPinBrickBrick,
            withPinBrickClear,
            withPinBrickRound,
            withPinClearBrick,
            withPinClearClear,
            withPinClearRound,
        ),
        withPinRoundBrick,
        withPinRoundClear,
        withPinRoundRound,
    ),
    composeU(
        withViewDefault,
        withViewMaterial,
    ),
    withBaseline,
    withHasClear,
)(TextinputDesktop);

export type ITextinputProps = ExtractProps<typeof Textinput>;
