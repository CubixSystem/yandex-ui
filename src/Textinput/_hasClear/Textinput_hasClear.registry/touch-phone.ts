import { TextinputClear } from '../../Clear/Textinput-Clear';
import { textinputRegistry as textinputRegistryBase } from '../../Textinput.registry/touch-phone';

export const textinputRegistry = textinputRegistryBase.set('Clear', TextinputClear);
