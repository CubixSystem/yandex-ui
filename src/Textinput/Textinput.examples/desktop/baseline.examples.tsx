import React from 'react';

import { Textinput } from '@yandex-lego/components/Textinput/desktop/bundle';

export const Baseline = () => (
    <div style={{ display: 'inline-block', width: 330 }}>
        <Textinput baseline size="m" view="default" defaultValue="Hello" style={{ margin: 4, width: 150 }} />
        <Textinput baseline size="s" view="default" defaultValue="World" style={{ margin: 4, width: 150 }} />
    </div>
);
