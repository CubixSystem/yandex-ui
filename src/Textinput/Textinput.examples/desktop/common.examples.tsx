import React, { useState, useCallback } from 'react';
import { action } from '@storybook/addon-actions';

import { Textinput } from '../../Textinput.bundle/desktop';
import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';
import { Spacer } from '../../../Spacer/Spacer';
import { Spin } from '../../../Spin/Spin.bundle/desktop';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: createDecorators({ scope: 'desktop' }),
    parameters,
};

function updateCaret(node: HTMLElement): void {
    if (typeof window.getSelection !== 'undefined' && typeof document.createRange !== 'undefined') {
        const range = document.createRange();
        range.selectNodeContents(node);
        range.collapse(false);
        const selection = window.getSelection();
        if (selection !== null) {
            selection.removeAllRanges();
            selection.addRange(range);
        }
    }
}

export const CustomControl = () => {
    const [value1, setValue1] = useState('тут внутри div contenteditable, и всё работает');

    const renderControl = useCallback(({ type, value, onChange, controlRef, ...props }: any) => {
        return (
            <div
                style={{ display: 'flex', alignItems: 'center', padding: 'var(--textarea-size-m-spaceAll)' }}
                ref={controlRef}
                contentEditable
                className="Textinput-Control"
                onInput={(event) => {
                    if (onChange) {
                        const value = (event.target as HTMLDivElement).textContent || '';
                        (event.target as HTMLInputElement).value = value;
                        (event.currentTarget as HTMLInputElement).value = value;
                        updateCaret(event.target as HTMLDivElement);
                        onChange(event);
                    }
                }}
                {...props}
            >
                {value}
            </div>
        );
    }, []);

    return (
        <Textinput
            hasClear
            size="m"
            view="default"
            value={value1}
            onChange={(event) => setValue1(event.target.value)}
            renderControl={renderControl}
        />
    );
};

export const Baseline = () => (
    <div style={{ display: 'inline-block', width: 330 }}>
        <Textinput baseline size="m" view="default" defaultValue="Hello" style={{ margin: 4, width: 150 }} />
        <Textinput baseline size="s" view="default" defaultValue="World" style={{ margin: 4, width: 150 }} />
    </div>
);

Baseline.story = {
    name: 'baseline',
};

export const HasClear = () => {
    const [value1, setValue1] = useState('');
    const [value2, setValue2] = useState('');

    return (
        <>
            <div style={{ padding: 4 }}>
                <Textinput
                    hasClear
                    size="m"
                    view="default"
                    value={value1}
                    onChange={(event) => setValue1(event.target.value)}
                    onClearClick={() => setValue1('')}
                />
            </div>
            <div style={{ padding: 4 }}>
                <Textinput
                    hasClear
                    size="m"
                    theme="normal"
                    value={value2}
                    onChange={(event) => setValue2(event.target.value)}
                    onClearClick={() => setValue2('')}
                />
            </div>
        </>
    );
};

HasClear.story = {
    name: 'hasClear',
};

export const Pin = () => {
    const rPins = ['round-round', 'round-clear', 'clear-round', 'round-brick', 'brick-round'];
    const qPins = ['brick-brick', 'brick-clear', 'clear-brick'];
    return (
        <>
            <p>Поля со скругленными уголками</p>
            {rPins.map((pin: any) => (
                <Textinput key={pin} size="m" view="default" pin={pin} value={`pin=${pin}`} />
            ))}
            <p>Поля с прямоугольными уголками</p>
            {qPins.map((pin: any) => (
                <Textinput key={pin} size="m" view="default" pin={pin} value={`pin=${pin}`} />
            ))}
            <p>Поле без боковых границ</p>
            <Textinput size="m" view="default" pin="clear-clear" value="pin=clear-clear" />
        </>
    );
};

Pin.story = {
    name: 'pin',
};

export const Size = () => (
    <>
        <div style={{ padding: 4 }}>
            <Textinput size="m" view="default" defaultValue="size m" />
        </div>
        <div style={{ padding: 4 }}>
            <Textinput size="s" view="default" defaultValue="size s" />
        </div>
    </>
);

Size.story = {
    name: 'size',
};

export const Theme = () => (
    <>
        <div style={{ padding: 4 }}>
            <Textinput theme="normal" size="m" defaultValue="theme normal" />
        </div>
        <div style={{ padding: 4 }}>
            <Textinput theme="websearch" defaultValue="theme websearch" />
        </div>
    </>
);

Theme.story = {
    name: 'theme',
};

export const Type = () => (
    <>
        <div style={{ padding: 4 }}>
            <Textinput size="m" view="default" type="number" defaultValue="200" />
        </div>
        <div style={{ padding: 4 }}>
            <Textinput size="m" view="default" type="password" defaultValue="secret" />
        </div>
    </>
);

Type.story = {
    name: 'type',
};

export const View = () => (
    <>
        <Textinput size="m" view="default" defaultValue="view default" />
    </>
);

View.story = {
    name: 'view',
};

export const State = () => (
    <>
        <Textinput state="error" hint="Error message" size="m" view="default" defaultValue="view default" />
    </>
);

State.story = {
    name: 'state',
};

export const Debounce = () => {
    const [value, setValue] = useState('');
    const [typing, setTyping] = useState(false);
    return (
        <Textinput
            debounceTimeout={500}
            size="m"
            view="default"
            value={value}
            style={{ maxWidth: 300 }}
            onChange={(event) => {
                action('onChange')(event.target.value);
                setTyping(false);
                setValue(event.target.value);
            }}
            onInput={(event) => {
                action('onInput')(event.target.value);
                setTyping(true);
            }}
            iconRight={(typing || undefined) && (
                <Spacer all="10px">
                    <Spin view="default" size="xxs" progress />
                </Spacer>
            )}
        />
    );
};

Debounce.story = {
    name: 'debounce',
};
