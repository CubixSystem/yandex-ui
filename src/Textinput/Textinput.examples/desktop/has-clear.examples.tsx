import React, { useState } from 'react';

import { Textinput } from '@yandex-lego/components/Textinput/desktop/bundle';

export const HasClear = () => {
    const [value1, setValue1] = useState('');
    const [value2, setValue2] = useState('');

    return (
        <>
            <div style={{ padding: 4 }}>
                <Textinput
                    hasClear
                    size="m"
                    view="default"
                    value={value1}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => setValue1(event.target.value)}
                    onClearClick={() => setValue1('')}
                />
            </div>
            <div style={{ padding: 4 }}>
                <Textinput
                    hasClear
                    size="m"
                    theme="normal"
                    value={value2}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => setValue2(event.target.value)}
                    onClearClick={() => setValue2('')}
                />
            </div>
        </>
    );
};
