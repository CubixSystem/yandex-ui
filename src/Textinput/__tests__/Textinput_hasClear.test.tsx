import React, { createRef } from 'react';

import { createClientRender, fireEvent } from '../../internal/testing';
import { Textinput as TextinputDesktop, withHasClear as withHasClearDesktop } from '../desktop';
import { Textinput as TextinputTouchPad, withHasClear as withHasClearTouchPad } from '../touch-pad';
import { Textinput as TextinputTouchPhone, withHasClear as withHasClearTouchPhone } from '../touch-phone';

const platforms = [
    ['desktop', withHasClearDesktop(TextinputDesktop)],
    ['touch-pad', withHasClearTouchPad(TextinputTouchPad)],
    ['touch-phone', withHasClearTouchPhone(TextinputTouchPhone)],
] as const;

describe.each(platforms)('Textinput_hasClear@%s', (_platform, Textinput) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Textinput hasClear />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен вызывать событие onChange с пустым значением в event при клике в clear', () => {
        const onChange = jest.fn();
        const { container } = render(<Textinput hasClear onChange={onChange} />);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        fireEvent.click(container.querySelector('.Textinput-Clear')!);

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange.mock.calls[0][0].target.value).toBe('');
    });

    test('должен вызывать событие onClearClick при клике в clear', () => {
        const onClearClick = jest.fn();
        const { container } = render(<Textinput hasClear onClearClick={onClearClick} />);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        fireEvent.click(container.querySelector('.Textinput-Clear')!);

        expect(onClearClick).toHaveBeenCalledTimes(1);
    });

    test('должен проставлять фокус в input после клика в clear', async () => {
        const controlRef = createRef<HTMLInputElement>();
        const { container } = render(<Textinput hasClear controlRef={controlRef} />);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        fireEvent.click(container.querySelector('.Textinput-Clear')!);

        expect(controlRef.current).toHaveFocus();
    });
});
