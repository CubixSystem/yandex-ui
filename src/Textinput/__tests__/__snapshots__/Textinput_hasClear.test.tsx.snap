// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textinput_hasClear@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
    value=""
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
    value=""
  />
  <span
    class="Textinput-Box"
  />
</span>
`;

exports[`Textinput_hasClear@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Textinput Textinput_hasClear"
>
  <span
    aria-hidden="true"
    class="Icon Icon_type_cross Textinput-Icon Textinput-Clear"
  />
  <input
    aria-invalid="false"
    autocomplete="off"
    class="Textinput-Control"
    value=""
  />
  <span
    class="Textinput-Box"
  />
</span>
`;
